core/tracker/tracker.go

```diff
@@ -134,11 +149,55 @@ func analyseDutyFailed(duty core.Duty, es []event) (bool, component, string) {
 // TODO(xenowits): Implement logic for reporting duties.
 func failedDutyReporter(core.Duty, bool, string, string) {}
 
-// analyseParticipation returns the share indexes of peers that participated in this duty.
-// TODO(dhruv): implement logic to analyse participation.
-//nolint:deadcode
-func analyseParticipation(core.Duty, []event) []int {
-	return nil
+// analyseParticipation returns a set of share indexes of participated peers corresponding to each DV public key.
+func analyseParticipation(events []event) map[core.PubKey]map[shareIdx]bool {
+	// Set of shareIdx of participated peers per DV.
+	resp := make(map[core.PubKey]map[shareIdx]bool)
+
+	eventsPerDV := make(map[core.PubKey][]event)
+	for _, e := range events {
+		eventsPerDV[e.pubkey] = append(eventsPerDV[e.pubkey], e)
+	}
+
+	for pubKey, dvEvents := range eventsPerDV {
+		resp[pubKey] = make(map[shareIdx]bool)
+		for _, e := range dvEvents {
+			// If we get a parSigDBInternal event, then the current node participated successfully.
+			// If we get a parSigEx event, then the corresponding peer with e.shareIdx participated successfully.
+			if e.component == parSigDBInternal || e.component == parSigEx {
+				resp[pubKey][e.shareIdx] = true
+			}
+		}
+	}
+
+	return resp
+}
+
+// newParticipationReporter returns a new participation reporter function which logs and instruments peer participation
+func newParticipationReporter(peers []p2p.Peer) func(context.Context, core.Duty, map[core.PubKey]map[shareIdx]bool, map[core.PubKey]map[shareIdx]bool) {
+	return func(ctx context.Context, duty core.Duty, currentParticipation map[core.PubKey]map[shareIdx]bool, lastParticipation map[core.PubKey]map[shareIdx]bool) {
+		for pubKey, dvPeers := range currentParticipation {
+			var absentPeers []string
+			for _, peer := range peers {
+				// Peer index is 0 indexed while shareIdx is 1 indexed.
+				if dvPeers[shareIdx(peer.Index+1)] {
+					participationGauge.WithLabelValues(duty.String(), peer.Name, pubKey.String()).Set(1)
+				} else {
+					absentPeers = append(absentPeers, peer.Name)
+					participationGauge.WithLabelValues(duty.String(), peer.Name, pubKey.String()).Set(0)
+				}
+			}
+
+			// Avoid spamming from identical logs.
+			if len(absentPeers) > 0 && !reflect.DeepEqual(currentParticipation[pubKey], lastParticipation[pubKey]) {
```

Comment:
best not use reflect in production code.

Also you should also log if some peers didn't participate last time, and all participated this time, e.g. "all peers participated". 

Note this is still going to spam the logs if we have 1000 validators in a cluster.